platforms/core-runtime/messaging/src/main/java/org/gradle/internal/event/AnonymousListenerBroadcast.java

```diff
@@ -29,8 +29,7 @@ public AnonymousListenerBroadcast(Class<T> type, Dispatch<MethodInvocation> forw
     }
 
     @Override
-    public void removeAll() {
-        super.removeAll();
-        add(forwardingDispatch);
+    public synchronized void removeAll() {
```

Comment:
💭 `synchronized` on this method should have been enough, but this is fine too.

Reply:
Hmm... I got it wrong then - I thought the fact we are reading without locking would allow the broadcast to be seen between these two mutating operations, leading to the problem you described. No?

Reply:
> Hmm... I got it wrong then - I thought the fact we are reading without locking would allow the broadcast to be seen between these two mutating operations, leading to the problem you described. No?
> 
> [#31692 (comment)](https://github.com/gradle/gradle/pull/31692#discussion_r1890134584)

Shouldn't write comments at 6am, I somehow thought that you've made reads synchronized too.